include/ssm.h

```diff
@@ -633,6 +380,215 @@ void ssm_desensitize(ssm_trigger_t *trig);
 
 /** @} */
 
+/**
+ * @addtogroup mem
+ * @{
+ */
+
+/** @brief Magic @a val_count value indicating a heap object is a builtin. */
+#define SSM_BUILTIN 0
+
+/** @brief Built-in types that are stored in the heap.
+ *
+ *  Type enumerated here are chosen because they cannot be easily or efficiently
+ *  expressed as a product of words. For instance, 64-bit timestamps cannot be
+ *  directly stored in the payload of a regular heap object, where even-numbered
+ *  timestamps may be misinterpreted as pointers.
+ */
+enum ssm_builtin {
+  SSM_TIME_T = 1, /**< 64-bit timestamps, #ssm_time_t */
+  SSM_SV_T,       /**< Scheduled variables, #ssm_sv_t */
+};
+
+/** @brief The metadata accompanying any heap-allocated object.
+ *
+ *  When @a val_count is 0, @a tag is to be interpreted as an #ssm_builtin.
+ */
+struct ssm_mm_header {
+  uint8_t val_count; /**< Number of ssm_value_t values in payload. */
+  uint8_t tag;       /**< Which variant is inhabited by this object. */
+  uint8_t ref_count; /**< The number of references to this object. */
+  uint32_t padding;  /**< @todo: get rid of this. */
```

Comment:
Wart: we need to explicitly pad this because otherwise the compiler would do this for us inside of a `struct ssm_mm` (only on 64-bit archs), throwing off size calculations.